wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,166 @@
+
+  static {
+    PropertyDefinition.registerPluginProperties(CustomEndpointPlugin.class);
+  }
+
+  protected final PluginService pluginService;
+  protected final RdsUtils rdsUtils = new RdsUtils();
+  protected final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc;
+  protected CustomEndpointMonitor customEndpointMonitor;
+
+  public CustomEndpointPlugin(final PluginService pluginService, final Properties props) {
+    this(
+        pluginService,
+        (hostSpec, region) ->
+            RdsClient.builder()
+                .region(region)
+                .credentialsProvider(AwsCredentialsManager.getProvider(hostSpec, props))
+                .build());
+  }
+
+  public CustomEndpointPlugin(
+      final PluginService pluginService,
+      final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc) {
+    this.pluginService = pluginService;
+    this.rdsClientFunc = rdsClientFunc;
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, isInitialConnection, connectFunc);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, isInitialConnection, forceConnectFunc);
+  }
+
+  protected Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      boolean isInitialConnection,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    final Connection conn = connectFunc.call();
+    if (!isInitialConnection || !this.rdsUtils.isRdsCustomClusterDns(hostSpec.getHost())) {
+      return conn;
+    }
+
+    // If we get here we are making an initial connection to a custom cluster URL
+    String customClusterHost = hostSpec.getHost();
+    CustomEndpointMonitor monitor = monitors.get(customClusterHost, MONITOR_EXPIRATION_NANO);
+    if (monitor != null) {
+      // A monitor is already running for this custom endpoint.
+      this.customEndpointMonitor = monitor;
+      monitor.registerPluginService(this.pluginService);
+      return conn;
+    }
+
+    this.customEndpointMonitor = monitors.computeIfAbsent(
+        customClusterHost,
+        (customEndpoint) -> new CustomEndpointMonitorImpl(
+            hostSpec,
+            props,
+            this.rdsClientFunc,
+            CUSTOM_ENDPOINT_INFO_EXPIRATION_NANO
+        ),
+        MONITOR_EXPIRATION_NANO
+    );
+
+    this.customEndpointMonitor.registerPluginService(this.pluginService);
```

Comment:
Cn we get Custom Endpoint status and update plugin service (instead of registering plugin service at monitor)?

Reply:
We could, which would allow connections to get initial custom endpoint info, but then there would be no way for the connections to update their custom endpoint info over time (at least not with the current implementation). The purpose of registering the plugin services is so that the custom endpoint monitor can update the plugin service with new custom endpoint info if the custom endpoint changes

Reply:
I'd still suggest to avoid registering plugin service at the monitor. Instead, the plugin can fetch custom endpoint info, check if there's any changes and update its local plugin service if needed.

Reply:
that is a better approach, but I'm not sure how to achieve it since the plugin service is not directly aware of which plugins are being used. With the host list providers, the plugin service has a direct reference to them so it can call `hostListProvider.refresh`. However, it can't call `customEndpointPlugin.refresh`. Any suggested solutions to this?

Reply:
After chatting with Sergiy and changing the design, we are using a single plugin service which can be used to updated the shared plugin service info cache with custom endpoint info.